### ISSUE TITLE
Modular c-code to interface with cython

### ISSUE THREAD

**Issue by smajee** [cabouman/cython_sandbox#15]
Currently parts of the c code (c structs and functions) need to be changed in order to interface with cython.
For a big code-base such as sv-mbirct or the cone-beam code, this would entail significant changes to the original code.

This can be simplified if we write a c-wrapper around the original c-code, to interface with cython.

I have implemented this in the branch `smajee`: https://github.com/cabouman/cython_sandbox/tree/smajee

Let me know what you think.

**Comment by tbalke** [comment]
YES!!

I like this a lot better.

According to Charlie's phrasing there is "crap" and we have to put it somewhere. Either in the python/cython part or in the c part. However, either case seems insufficient: the python part is what the user is interacting with, i.e. what is more visible from the outside, and modifying the c-code is ugly as it requires to modify/pollute a working module. This creative option here adds an intermediate layer (of crap) in the middle which seems to avoid the worst of both worlds.

- - -

A side note:
Since we have a multialloc function that works with N dimensions (arbitrary N), maybe it would be useful to have a `multiarray_from_flattened` routine that also operates from N dimensions.

This way a template for converting the structures would be simply:

```
void convert_foo(struct foo *s, struct foo_flat *s_flat)
{
  s->param1 = s_flat->param1;
  ...
  s->paramK = s_flat->paramK;
  multiarray_from_flattened(s->array, s_flat->array, 3, N1, N2, N3);
}
```
In other words, a recipe for conversion is: copy all scalers and `multiarray_from_flattened` all arrays.

**Comment by cabouman** [comment]
I personally feel the "crap" should go into the C world. (C is a more natural place for crap.) Someone can always write a C wrapper for their C function that does the conversion. 

I like the idea of ``multiarray_from_flattened()`` and ``flattened_from_multiarray()`` functions. But I'm not yet convince it's worth the effort yet.
